pkg/xray/reports.go

```diff
@@ -0,0 +1,802 @@
+}
+
+func unpackStartAndEndDate(d *schema.Set) *StartAndEndDate {
+	var dates StartAndEndDate
+
+	if len(d.List()) > 0 {
+		f := d.List()[0].(map[string]interface{})
+		dates = StartAndEndDate{
+			Start: f["start"].(string),
+			End:   f["end"].(string),
+		}
+		return &dates
+	}
+
+	return nil
+}
+
+func resourceXrayVulnerabilitiesReportCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	return createReport("vulnerabilities", d, m)
+}
+
+func resourceXrayLicensesReportCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	return createReport("licenses", d, m)
+}
+
+func resourceXrayViolationsReportCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	return createReport("violations", d, m)
+}
+
+func resourceXrayOperationalRisksReportCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	return createReport("operationalRisks", d, m)
+}
+
+func resourceXrayReportRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	report := Report{}
+
+	projectKey := d.Get("project_key").(string)
+	req, err := getRestyRequest(m.(*resty.Client), projectKey)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp, err := req.
+		SetResult(&report).
+		SetPathParams(map[string]string{
+			"reportId": d.Id(),
+		}).
+		Get("xray/api/v1/reports/{reportId}")
+	if err != nil {
+		if resp != nil && resp.StatusCode() == http.StatusNotFound {
+			tflog.Warn(ctx, fmt.Sprintf("Xray report (%s) not found, removing from state", d.Id()))
+			d.SetId("")
+		}
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceXrayReportDelete(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	report := unpackReportProjectKey(d)
+
+	req, err := getRestyRequest(m.(*resty.Client), report.ProjectKey)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	resp, err := req.
+		SetPathParams(map[string]string{
+			"reportId": d.Id(),
+		}).
+		Delete("xray/api/v1/reports/{reportId}")
+	if err != nil && resp.StatusCode() == http.StatusNotFound {
+		d.SetId("")
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func createReport(reportType string, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	report := unpackReport(d, reportType)
+	req, err := getRestyRequest(m.(*resty.Client), report.ProjectKey)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	_, err = req.SetBody(report).SetResult(&report).
+		Post("xray/api/v1/reports/" + reportType)
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%d", report.ReportId))
+
+	return nil
+}
+
+func reportResourceDiff(_ context.Context, diff *schema.ResourceDiff, v interface{}) error {
+	reportResources := diff.Get("resources").(*schema.Set).List()
+	if len(reportResources) == 0 {
+		return nil
+	}
+
+	// Verify only one resource attribute is set.
+	for _, reportResource := range reportResources {
```

Comment:
Just a comment. These ugly lines won't be necessary when we migrate to the new Plugin Framework 😄